File: 92.reverse-linked-list-ii.py
#
# @lc app=leetcode id=92 lang=python3
#
# [92] Reverse Linked List II
#

# Time:  O(n)
# Space: O(1)


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        dummy = ListNode(-1, head)
        prev = dummy
        for i in range(left - 1):
            prev = prev.next
        curr = prev.next
        for i in range(left, right):
            next = curr.next
            curr.next = next.next
            next.next = prev.next
            prev.next = next
        return dummy.next


# @lc code=end

File: 92.reverse-linked-list-ii.js
/*
 * @lc app=leetcode id=92 lang=javascript
 *
 * [92] Reverse Linked List II
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function (head, left, right) {
  const dummy = new ListNode(-1, head);
  let prev = dummy;
  for (let i = 0; i < left - 1; ++i) {
    prev = prev.next;
  }
  let curr = prev.next;
  for (let i = left; i < right; ++i) {
    const next = curr.next;
    curr.next = next.next;
    next.next = prev.next;
    prev.next = next;
  }
  return dummy.next;
};
// @lc code=end
